### DISCUSSION TITLE
Handling Multi-Channel Inputs

### DISCUSSION THREAD

**Discussion by vishjain** [espnet/espnet#5704] [Q&A]
Hi, 

I've noticed many of the speech enhancement algorithms have multiple channels (5 or 10 even). 

Some of my audio data when processed from soundfile/librosa has 2 channels. However, the # of channels isn't equivalent to the expected input of the multi-channel speech enhancement model.

What are the canonical ways of expanding the data to more channels without having to train or build out a new model? 

For example, I've tried adding 0s to the extra channels or copied and pasted the data in the original channels but that hasn't worked too well.

**Comment by sw005320** [comment]
Thanks for your email.
This is a very good question.
It actually depends on the algorithm.
Some multichannel speech enhancement models assume the specific microphone geometry during training, and we cannot use the other multichannel signals during inference.
However, mask-based MVDR or microphone-geometry-invariant methods are robust against it.
It can also be used in cases where we have different microphones.

@Emrys365, can you follow up on this discussion?
Can Your USES be used for this case?

**Comment by Emrys365** [comment]
Sorry for the late response.

In most existing NN-based multichannel speech enhancement (SE) models, it is likely that they learned to process a fixed array geometry depending on the training data. For example, CHiME-4 data features a 6-mic rectangular array.
In some other cases, the training data may consist of several microphone arrays. Then the model may be able to process a wider range of array geometries. However, it still requires the input microphone channels to contain actual signals with a consistent spatial correlation. So I would expect that simply concatenating an all-zero channel or copying one channel to another will not work for such systems.

Instead, there are three major solutions.
1. Only use a single channel as input, and use any single-channel SE model. This is of course sub-optimal and requires selection of the microphone channel.
2. Use a single-channel SE model along with a subsequent beamformer (e.g., MVDR). The SE model's output can be used to [estimate masks for performing multichannel beamforming](https://github.com/espnet/espnet/blob/master/espnet2/enh/layers/dnn_beamformer.py#L94), which usually improves the perceptual quality and downstream performance. But also note that beamforming does not fully suppress the inteference/noise due to its distortionless constraint.
3. Use an array-geometry-agnostic NN-based SE model. Some NN-based SE models are designed to process any microphone array or number. They adopt techniques such as Transform-Average-Concatenate (TAC), channel-wise attention, and so on to allow this type of processing. They are usually trained on data with different array geometries, and can generalize to unseen geometries. One off-the-shelf array-geometry-agnostic SE model is available in ESPnet, called [USES](https://arxiv.org/abs/2309.17384). You can follow the link below to enhance audios with this model: https://huggingface.co/espnet/Wangyou_Zhang_universal_train_enh_uses_refch0_2mem_raw
